service/controller/v14/resource/bridgezone/resource.go

```diff
@@ -0,0 +1,415 @@
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "intermediate zone found")
+	}
+
+	var finalZoneID string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "searching for final zone in guest account")
+
+		finalZoneID, err = r.findHostedZoneID(ctx, defaultGuest, finalZone)
+		if IsNotFound(err) {
+			// The final zone is not yet created. Retry in the next
+			// reconciliation loop.
+			r.logger.LogCtx(ctx, "level", "debug", "message", "final zone not found, skipping the resource")
+			return nil
+		} else if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "final zone found")
+	}
+
+	var finalZoneRecords []*route53.ResourceRecord
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "getting final zone name servers and TTL from intermediate")
+
+		nameServers, _, err := r.getNameServersAndTTL(ctx, guest, finalZoneID, finalZone)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		for _, ns := range nameServers {
+			copy := ns
+			v := &route53.ResourceRecord{
+				Value: &copy,
+			}
+			finalZoneRecords = append(finalZoneRecords, v)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "got final zone name servers and TTL from intermediate zone")
+	}
+
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring intermediate zone delegation")
+
+		upsert := route53.ChangeActionUpsert
+		ns := route53.RRTypeNs
+		ttl := int64(900)
+
+		in := &route53.ChangeResourceRecordSetsInput{
+			ChangeBatch: &route53.ChangeBatch{
+				Changes: []*route53.Change{
+					{
+						Action: &upsert,
+						ResourceRecordSet: &route53.ResourceRecordSet{
+							Name:            &finalZone,
+							Type:            &ns,
+							TTL:             &ttl,
+							ResourceRecords: finalZoneRecords,
+						},
+					},
+				},
+			},
+			HostedZoneId: &intermediateZoneID,
+		}
+		_, err := defaultGuest.ChangeResourceRecordSetsWithContext(ctx, in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "ensured intermediate zone delegation")
+	}
+
+	return nil
+}
+
+func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
+	customObject, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !r.route53Enabled {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "route53 disabled, skipping execution")
+		return nil
+	}
+
+	baseDomain := key.BaseDomain(customObject)
+	intermediateZone := "k8s." + baseDomain
+	finalZone := key.ClusterID(customObject) + ".k8s." + baseDomain
+
+	_, defaultGuest, err := r.route53Clients(ctx)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var intermediateZoneID string
+	{
+		r.logger.LogCtx(ctx, "level", "debug", "message", "searching for intermediate zone in default guest account")
+
+		intermediateZoneID, err = r.findHostedZoneID(ctx, defaultGuest, intermediateZone)
+		if IsNotFound(err) {
```

Comment:
When can this actually happen and how can we still ensure the cleanup of certain resources or configurations?

Reply:
This can happen when we have a brand new installation or when we are after the migration period.
In general this resource does changes only to the intermediate zone, so if it doesn't exist we don't have to care about the deletion.

Reply:
I also wrote in the Resource godoc that when intermediate doesn't exist this resource becomes noop.